osquery/tables/system/windows/groups.cpp

```diff
@@ -0,0 +1,121 @@
+// clang-format on
+
+#include <osquery/core.h>
+#include <osquery/tables.h>
+#include <osquery/logger.h>
+
+#include "osquery/core/process.h"
+#include "osquery/core/windows/wmi.h"
+#include "osquery/tables/system/windows/registry.h"
+#include "osquery/core/conversions.h"
+
+namespace osquery {
+  
+  // Defined in /osquery/core/windows/process_ops.cpp, but not declared in a header:
+  std::string psidToString(PSID sid);
+  
+  // Get the relative identifier (RID) from a security identifier (SID):
+  unsigned long getRidFromSid(PSID sidPtr) {
+    BYTE* countPtr = GetSidSubAuthorityCount(sidPtr);
+    unsigned long indexOfRid = static_cast<unsigned long>(*countPtr - 1);
+    unsigned long* ridPtr = GetSidSubAuthority(sidPtr, indexOfRid);
+    return *ridPtr;
+  }
+
+  namespace tables {
+
+    std::unique_ptr<BYTE[]> GetSid(LPCWSTR accountName)
+    {
+      // Validate the input parameters.
+      if (accountName == nullptr)
+      {
+        LOG(WARNING) << "GetSid(): no account name provided.";
+        return nullptr;
+      }
+      
+      // Call LookupAccountNameW() once to retrieve the necessary buffer sizes for the SID (in bytes) and the domain name (in TCHARS):
+      unsigned long sidBufferSize = 0;
+      unsigned long domainNameSize = 0;
+      auto eSidType = SidTypeUnknown;
+      LookupAccountNameW(nullptr, accountName, nullptr, &sidBufferSize, nullptr, &domainNameSize, &eSidType);
+
+      // Allocate sufficient buffers for the (binary data) SID and the (wide string) domain name:
+      auto sidBuffer = std::make_unique<BYTE[]>(sidBufferSize);
+      std::vector<wchar_t> domainName(domainNameSize);
+
+      // Call LookupAccountNameW() a second time to actually obtain the SID for the given account name:
+      auto ret = LookupAccountNameW(nullptr, accountName, sidBuffer.get(), &sidBufferSize, domainName.data(), &domainNameSize, &eSidType);
+      if (ret == 0) {
+        LOG(WARNING) << "Failed to LookupAccountNameW(): " << accountName;
+      }
+      else if (IsValidSid(sidBuffer.get()) == FALSE) {
+        LOG(WARNING) << "The SID for " << accountName << " is invalid.";
+      }
+
+      return sidBuffer; // Implicit move operation. Caller now "owns" the returned pointer.
+    }
+
+    void processLocalGroups(QueryData& results) {
+      unsigned long groupInfoLevel = 1;
+      unsigned long numGroupsRead = 0;
+      unsigned long totalGroups = 0;
+      unsigned long resumeHandle = 0;
+      unsigned long ret = 0;
+      LOCALGROUP_INFO_1 *lginfo = nullptr;
+      
+      std::unique_ptr<BYTE[]> sidSmartPtr = nullptr;
+      PSID sidPtr = nullptr;
+
+      do {
+        ret = NetLocalGroupEnum(nullptr, groupInfoLevel, (LPBYTE *)&lginfo, MAX_PREFERRED_LENGTH, &numGroupsRead, &totalGroups, nullptr);
+
+        if ((ret == NERR_Success || ret == ERROR_MORE_DATA) && lginfo != nullptr) {
+          for (size_t i = 0; i < numGroupsRead; i++) {
+            Row r;
+            sidSmartPtr = GetSid(lginfo[i].lgrpi1_name);

```

Comment:
Couldn't you have returned a `nullptr` here?

Reply:
If `ret == nullptr` then we would hit the `else` case below and log a warning. Are you referring to a different variable?